CDAP-19050 remove GCE calls when they are not needed

albertshau opened the pull request:
Refactored DataprocClient to use a new NetworkClient interface
to perform all operations that used to require a call to the
GCE APIs. A ComputeNetworkClient implementation contains all the
existing logic, while a NoOpNetworkClient implementation performs
all operations without making any external calls. When SSH is
required to monitor the program run, the ComputeNetworkClient is
used. When runtime monitoring is used, the NoOpNetworkClient is
used.

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+  @Override
+  public void init(@Nullable String networkName) throws RetryableProvisionException, IOException,
+    GeneralSecurityException {
+    compute = getCompute(conf);
```
this content was copied from the latter half of DataprocClient.fromConf()

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+  @Override
+  public Node getNode(Node.Type type, String zone, String nodeName) throws IOException {
+    Instance instance;
+    try {
```
this method was moved from DataprocClient.getNode()

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+  }
+
+  @Override
+  public List<String> getAdditionalTags(boolean isInternalIPOnly) throws RetryableProvisionException, IOException {
```
this method was moved from DataprocClient.getFirewallTags()

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
-    String subnet = conf.getSubnet();
-    Network networkInfo = getNetworkInfo(networkHostProjectId, network, compute);
+    NetworkClient networkClient = requireNetwork ? new ComputeNetworkClient(conf) : new NoOpNetworkClient();
+    networkClient.init(network);
```
the logic removed from this method was put into ComputeNetworkClient.init()

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
-    return new DataprocClient(new DataprocConf(conf, network, subnet), client, compute, networkInfo);
-  }
-
-  private static PeeringState getPeeringState(String systemProjectId, String systemNetwork, Network networkInfo) {
```
everything else removed from this class was moved to ComputeNetworkClient without any changes.

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
-    } else if (network == null) {
-      // Otherwise, pick a network from the configured project using the Compute API
-
-      network = findNetwork(networkHostProjectId, compute);
```
this is the main functional change. This logic still happens in ComputeNetworkClient.init(), but not in NoOpNetworkClient(). NoOpNetworkClient is used when runtime monitoring is being used (in a GCP environment) instead of SSH monitoring (in non-GCP environment). 

If the dataproc network is not given, and the network cannot be auto-detected, we now throw an exception saying the network must be provided instead of making a GCE call to pick a network.

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+/**
+ * A NetworkClient that is backed by actual calls to the GCP Compute APIs.
+ */
+public class ComputeNetworkClient implements NetworkClient {
```
basically the entirety of this class is moved from DataprocClient

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+  }
+
+  @Override
+  public boolean isNetworkPeered(String systemProjectId, String systemNetwork) {
```
this is one of the few parts that is not cut and paste from DataprocClient. This is taken from the end of DataprocClient.isInternalIPOnly()

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,442 @@
+  }
+
+  @Override
+  public void init(@Nullable String networkName) throws RetryableProvisionException, IOException,
```
this is taken from the latter half of DataprocClient.fromConf()

fernst commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java:
```diff
@@ -105,10 +106,11 @@ public static void main(String[] args) throws Exception {
     }
 
-    String imageVersion = commandLine.hasOption('i') ? commandLine.getOptionValue('i') : "1.2";
+    String imageVersion = commandLine.hasOption('i') ? commandLine.getOptionValue('i') : "1.3";
```
Why don't we update this default to 2.0? 1.3 is currently unsupported by Dataproc.

fernst reviewed (commented):
Asked a question. Other than that, code looks good.

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:
```diff
@@ -0,0 +1,441 @@
+  }
+
+  @Override
+  public boolean isNetworkPeered(String systemProjectId, String systemNetwork) {
```
Please add javadocs for all public methods

albertshau replied:
have added docs to the interface

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -103,73 +78,23 @@
-      }
-      throw e;
-    }
+  DataprocClient(DataprocConf conf, ClusterControllerClient client, NetworkClient nodeClient) {
```
Given factory method below, should this be private?

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -436,22 +212,21 @@ ClusterOperationMetadata createCluster(String name, String imageVersion, Map<Str
         clusterConfig.setSubnetworkUri(conf.getSubnet());
       } else {
-        clusterConfig.setNetworkUri(network.getSelfLink());
+        clusterConfig.setNetworkUri(nodeClient.getNetwork());
```
Any specific reason to name it `nodeClient` here and `networkClient` in `fromConf`?

tivv replied:
Also as there are cases where we don't fill `network`, could you add assert on any `getNetwork` call to ensure we don't try to use it when it's not available?

albertshau replied:
yep, this was an old name I forgot to change, will make it consistent.

This will always be non-null, will add a comment on the relevant part

tivv reviewed (commented):
So, my high level question is: We use network name to create a cluster. Does it mean this PR changes it to null if we are not using SSH?

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java:
```diff
@@ -57,6 +57,7 @@ public static void main(String[] args) throws Exception {
       .addOption(new Option("p", "project", true, "Google cloud project id."))
       .addOption(new Option("c", "configFile", true, "File all provisioner settings as a json object."))
       .addOption(new Option("i", "imageVersion", true, "The image version for the cluster. Defaults to 1.2."))
```
Change default here are you are changing it below. And I agree that 2.0 should be default

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java:
```diff
@@ -419,7 +418,7 @@ protected String getClusterName(ProvisionerContext context) throws Exception {
   protected DataprocClient getClient(DataprocConf conf) throws IOException, GeneralSecurityException,
     RetryableProvisionException {
-    return DataprocClient.fromConf(conf);
+    return DataprocClient.fromConf(conf, conf.getPublicKey() != null);
```
It would be great not to do those calls for deprovisioning as it's not needed irrespective of SSH or not

albertshau replied:
good point, I'll make this change.

albertshau commented:
Sorry I forgot to push latest changes, the network (for creation) is supposed to always be non-null. We still need to look it up when it is not given, even without SSH.

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -180,145 +105,31 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
-        return currentSubnet;
-      }
+      // Otherwise, make some Compute calls to pick a network
+      networkClient.init(null);
```
over here, we still look up the network if null, even if SSH is not required.

I think I'll refactor this a bit, since we don't need to do this lookup for non-create operations.

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:
```diff
@@ -1127,7 +710,7 @@ private io.cdap.cdap.runtime.spi.provisioner.ClusterStatus convertStatus(Cluster
 
   @Override
-  public void close() {
+  public void close() throws Exception {
```
Was there `ClusterControllerClient` change?

tivv commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/Clients.java:
```diff
@@ -0,0 +1,113 @@
+
+    networkClient = shouldUseSSH ?
+      new ComputeClient(conf, Clients.getCompute(conf)) : new NoOpClient();
+    networkClient.init(network, subnet);
```
nit: combine this ternary with previous if. It would be easier to follow the logic.

tivv commented on cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisionerTest.java:
```diff
@@ -288,11 +291,17 @@ public void testClusterCreateNoReuse() throws Exception {
+    ComputeClient computeClientMock = Mockito.mock(ComputeClient.class);
+    PowerMockito.spy(Clients.class);
+    PowerMockito.doReturn(computeClientMock)
+      .when(Clients.class, "createNetworkClient", Mockito.any(), Mockito.eq(true));
```
nit: Instead of mocking a static method it's often a good idea to use a factory class instead of factory method

tivv reviewed (commented):
As of now the logic is pretty hard to follow, primarily because ComputeClient is doing two things: network detection & ssh-specific configuration data retrieval. Also ssh-specific part is split between DataprocClient  and NetworkClient and it's non-obvious of what would be  available for DataprocClient from NetworkClient depending on if SSH is needed or not.
I'd suggest to separate out network information / detection and ssh initialization into two separate entities (either unrelated or with inheritance). May be even make lazy network detection on on-needed basis. This should simplify a lot of "if else" code and organize all "if ssh" things in one place, making it much easier to comprehend

albertshau commented:
I'll see what I can do. One of the difficulties is the result of the initial call to get the network information is used in all the methods

tivv commented:
@albertshau how about doing it lazy on request? I think it would simplify things a lot. You can have just 1 implementation that just don't do this calls if they are not requested

albertshau committed:
CDAP-19050 refactor dataproc client to remove unneeded calls

Refactor the DataprocClient so that it only looks up network
information in the create cluster call, and it skips getting node
IPs and create times when SSH access is not needed.

Split the existing DataprocClient into two separate classes with
different logic for setting network information and getting node
information.

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/SSHDataprocClient.java:
```diff
@@ -0,0 +1,414 @@
+ * Wrapper around the dataproc client that adheres to our configuration settings. Creates Dataproc clusters that
+ * are accessible through SSH.
+ */
+class SSHDataprocClient extends DataprocClient {
```
this class is just code moved from DataprocClient verbatim

albertshau commented on cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/RuntimeMonitorDataprocClient.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Wrapper around the dataproc client that adheres to our configuration settings.
+ */
+class RuntimeMonitorDataprocClient extends DataprocClient {
```
behavior changes are local to this class, where it doesn't lookup node information and it doesn't look at networking setup to determine whether the cluster should be private ip only

tivv commented on cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClientTest.java:
```diff
@@ -61,13 +63,10 @@ public void init() throws Exception {
-    dataprocConf = DataprocConf.create(properties, null);
+    dataprocConf = DataprocConf.create(properties);
 
-    PowerMockito.spy(DataprocClient.class);
```
Thanks! static method mocking is always a code smell

albertshau replied:
👍  I didn't even know you could mock static methods until seeing this test
